actix-rt/src/system.rs

```diff
@@ -71,6 +119,76 @@ impl System {
             .run_nonblocking()
     }
 
+    /// Consume the provided tokio Runtime and start the `System` in it.
+    /// This method will create a `LocalSet` object and occupy the current thread
+    /// for the created `System` exclusively. All the other asynchronous tasks that
+    /// should be executed as well must be aggregated into one future, provided as the last
+    /// argument to this method.
+    ///
+    /// Note: This method uses provided `Runtime` to create a `System` future only.
+    /// All the [`Arbiter`]s will be started in separate threads using their own tokio `Runtime`s.
+    /// It means that using this method currently it is impossible to make `actix-rt` work in the
+    /// `Runtime` other than the provided by `tokio` 0.2 (e.g. provided by `tokio_compat`).
+    ///
+    /// [`Arbiter`]: struct.Arbiter.html
+    ///
+    /// # Arguments
+    ///
+    /// - `runtime`: A tokio Runtime to run the system in.
+    /// - `name`: Name of the System
+    /// - `rest_operations`: A future to be executed in the runtime along with the System.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use tokio::runtime::Runtime;
+    /// use actix_rt::System;
+    /// use futures::future::try_join_all;
+    ///
+    /// async fn run_application() {
+    ///     let first_task = tokio::spawn(async {
+    ///         // ...
+    /// #        println!("One task");
+    /// #        Ok::<(),()>(())
+    ///     });
+    ///
+    ///     let second_task = tokio::spawn(async {
+    ///         // ...
+    /// #       println!("Another task");
+    /// #       Ok::<(),()>(())
+    ///     });
+    ///
+    ///     try_join_all(vec![first_task, second_task])
+    ///         .await
+    ///         .expect("Some of the futures finished unexpectedly");
+    /// }
+    ///
+    ///
+    /// let runtime = tokio::runtime::Builder::new()
+    ///     .core_threads(2)
+    ///     .enable_all()
+    ///     .threaded_scheduler()
+    ///     .build()
+    ///     .unwrap();
+    ///
+    /// let rest_operations = run_application();
+    /// System::attach_to_tokio(runtime, "actix-main-system", rest_operations);
+    /// ```
+    pub fn attach_to_tokio<Fut, R>(
```

Comment:
I like the idea of this function, making it much simpler to host actix in an existing tokio runtime.

Are there alternate designs for the `rest_operations` idea worth discussing? I seem to recall in the past just `join`ing and awaiting the actix LocalSet with any other long running tasks. This approach feels more opinionated and less extensible.

Reply:
Well, I've attempted several other approaches, but found them not as convenient because of the following factors:
- `LocalSet` occupies current thread for `System` to run;
- Future, obtained via `LocalSet` is not `Send`. 

Thus, returning a future from this function (instead of blocking the runtime inside) makes the user-side code not really intuitive.

As an alternative approach, we may make `rest_operations` a function rather than a future, but IMHO it's less flexible: if you have an `async` function, you can get a future by simply calling it, and at the same time you can get a future without an additional function (e.g. using an `async { }` block).

Or do you have an idea how to get rid of the `rest_operations` argument without complications on the caller site?

Reply:
This feels like we're moving into a simpler Node.js style `<req, res>` or Koa's `ctx` and I'm all for it. I'd love to have Actix just be the web layer on top of default Tokio.

Reply:
I think this is similar to what @fafhrd91 was describing too about the `actix-net` features.